glean-core/csharp/csharp/LibGleanFFI.cs

```diff
@@ -25,6 +25,57 @@ internal class FfiConfiguration
             public bool delay_ping_lifetime_io;
         }
 
+        /// <summary>
+        /// A base handle class meant to be extended by the different metric types to allow
+        /// for calling metric specific clearing functions.
+        /// </summary>
+        internal class BaseGleanHandle : SafeHandle
+        {
+            public BaseGleanHandle() : base(invalidHandleValue: IntPtr.Zero, ownsHandle: true) { }
+
+            public override bool IsInvalid
+            {
+                get { return this.handle == IntPtr.Zero; }
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                // Note: this is meant to be implemented by the inheriting class in order to
+                // provide a specific cleanup action.
+                return false;
+            }
+        }
+
+        internal class StringAsReturnValue : SafeHandle
+        {
+            public StringAsReturnValue() : base(IntPtr.Zero, true) { }
+
+            public override bool IsInvalid
+            {
+                get { return this.handle == IntPtr.Zero; }
+            }
+
+            public string AsString()
+            {
+                int len = 0;
+                while (Marshal.ReadByte(handle, len) != 0) { ++len; }
```

Comment:
it is not clear to me how you get this `handle`, and it is the same with `this.handle` from other functions?

Reply:
The `SafeHandle` integrates with the C# P/Invoke, so can be safely used as a input or output (or return!) value. The framework will take care of passing things properly, additionally providing a GC-friendly way to clean up memory. See the the [official docs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle?view=netcore-3.1), [this nice post](https://www.meziantou.net/stop-using-intptr-for-dealing-with-system-handles.htm) and how the [FFI Omnibus](http://jakegoulding.com/rust-ffi-omnibus/string_return/) talks about it.